openmls/src/framing/mls_auth_content_in.rs

```diff
@@ -0,0 +1,356 @@
+                match &self.tbs.content.body {
+                    FramedContentBodyIn::Proposal(ProposalIn::Add(add_proposal)) => {
+                        Ok(add_proposal.unverified_credential())
+                    }
+                    _ => Err(ValidationError::NotAnExternalAddProposal),
+                }
+            }
+            _ => Err(ValidationError::UnknownMember),
+        }
+    }
+
+    /// Get the wire format.
+    pub(crate) fn wire_format(&self) -> WireFormat {
+        self.tbs.wire_format
+    }
+
+    /// Get the confirmation tag.
+    pub(crate) fn confirmation_tag(&self) -> Option<&ConfirmationTag> {
+        self.auth.confirmation_tag.as_ref()
+    }
+
+    /// Get the content type
+    pub(crate) fn content_type(&self) -> ContentType {
+        self.tbs.content.body.content_type()
+    }
+}
+
+impl Verifiable for VerifiableAuthenticatedContentIn {
+    fn unsigned_payload(&self) -> Result<Vec<u8>, tls_codec::Error> {
+        self.tbs.tls_serialize_detached()
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.auth.signature
+    }
+
+    fn label(&self) -> &str {
+        "FramedContentTBS"
+    }
+}
+
+impl VerifiedStruct<VerifiableAuthenticatedContentIn> for AuthenticatedContentIn {
+    fn from_verifiable(v: VerifiableAuthenticatedContentIn, _seal: Self::SealingType) -> Self {
+        AuthenticatedContentIn {
+            wire_format: v.tbs.wire_format,
+            content: v.tbs.content,
+            auth: v.auth,
+        }
+    }
+
+    type SealingType = private_mod::Seal;
+}
+
+impl SignedStruct<FramedContentTbsIn> for AuthenticatedContentIn {
+    fn from_payload(tbs: FramedContentTbsIn, signature: Signature) -> Self {
+        let auth = FramedContentAuthDataIn {
+            signature,
+            // Tags must always be added after the signature
+            confirmation_tag: None,
+        };
+        Self {
+            wire_format: tbs.wire_format,
+            content: tbs.content,
+            auth,
+        }
+    }
+}
+
+impl Size for FramedContentAuthDataIn {
+    #[inline]
+    fn tls_serialized_len(&self) -> usize {
+        self.signature.tls_serialized_len()
+            + if let Some(confirmation_tag) = &self.confirmation_tag {
+                confirmation_tag.tls_serialized_len()
+            } else {
+                0
+            }
+    }
+}
+
+impl TlsSerializeTrait for FramedContentAuthDataIn {
+    fn tls_serialize<W: Write>(&self, writer: &mut W) -> Result<usize, tls_codec::Error> {
+        let mut written = self.signature.tls_serialize(writer)?;
+        written += if let Some(confirmation_tag) = &self.confirmation_tag {
+            confirmation_tag.tls_serialize(writer)?
+        } else {
+            0
+        };
+        Ok(written)
+    }
+}
+
+// The following two `From` implementations break abstraction layers and MUST
+// NOT be made available outside of tests or "test-utils".
```

Comment:
Then there should probably be cfg statements for both of them, right? If not, can you move the TODO below up here?

Reply:
See below